packages/cloudflare/src/api/cloudflare-context.ts

```diff
@@ -0,0 +1,149 @@
+ *
+ * @returns the local proxy version of the cloudflare context
+ */
+async function getCloudflareContextInNextDev<
+  CfProperties extends Record<string, unknown> = IncomingRequestCfProperties,
+  Context = ExecutionContext,
+>(): Promise<CloudflareContext<CfProperties, Context>> {
+  const global = globalThis as unknown as {
+    [cloudflareContextInNextDevSymbol]: CloudflareContext<CfProperties, Context> | undefined;
+  };
+
+  if (!global[cloudflareContextInNextDevSymbol]) {
+    try {
+      const cloudflareContext = await getCloudflareContextFromWrangler<CfProperties, Context>();
+      global[cloudflareContextInNextDevSymbol] = cloudflareContext;
+    } catch (e: unknown) {
+      if (e instanceof Error && e.message.includes("A dynamic import callback was not specified")) {
+        const getCloudflareContextFunctionName = getCloudflareContext.name;
+        const enablingFunctionName = enableEdgeDevGetCloudflareContext.name;
+        throw new Error(
+          `\n\n\`${getCloudflareContextFunctionName}\` has been invoked during development inside the edge runtime ` +
+            "this is not enabled, to enable such use of the function you need to import and call the " +
+            `\`${enablingFunctionName}\` function inside your Next.js config file\n\n"` +
+            "Example: \n   ```\n   // next.config.mjs\n\n" +
+            `   import { ${enablingFunctionName} } from "@opennextjs/cloudflare";\n\n` +
+            `   ${enablingFunctionName}();\n\n` +
+            "   /** @type {import('next').NextConfig} */\n" +
+            "   const nextConfig = {};\n" +
+            "   export default nextConfig;\n" +
+            "   ```\n" +
+            "\n(note: currently middlewares in Next.js are always run using the edge runtime)\n\n"
+        );
+      } else {
+        throw e;
+      }
+    }
+  }
+
+  return global[cloudflareContextInNextDevSymbol]!;
+}
+
+type RuntimeContext = Record<string, unknown> & {
+  process?: { env?: Record<string | symbol, unknown> };
+  [cloudflareContextInNextDevSymbol]?: {
+    env: unknown;
+    ctx: unknown;
+    cf: unknown;
+  };
+};
+
+/**
+ * Enables `getCloudflareContext` to work, during development (via `next dev`), in the edge runtime
+ *
+ * Note: currently middlewares always run in the edge runtime
+ *
+ * Note: this function should only be called inside the Next.js config file
+ */
+export async function enableEdgeDevGetCloudflareContext() {
+  const require = (
+    await import(/* webpackIgnore: true */ `${"__module".replaceAll("_", "")}`)
+  ).default.createRequire(import.meta.url);
+
+  // eslint-disable-next-line unicorn/prefer-node-protocol -- the `next dev` compiler doesn't accept the node prefix
+  const vmModule = require("vm");
+
+  const originalRunInContext = vmModule.runInContext.bind(vmModule);
+
+  const context = await getCloudflareContextFromWrangler();
+
+  vmModule.runInContext = (...args: [string, RuntimeContext, ...unknown[]]) => {
```

Comment:
Extract this to a distinct function to add comments?
ref: https://github.com/cloudflare/next-on-pages/blob/2cd4c3c704a00e6b693229f1f14102abc6318d11/internal-packages/next-dev/src/shared.ts#L22